runtime/src/main/java/org/corfudb/runtime/view/ManagementView.java

```diff
@@ -319,125 +238,313 @@ private NodeStatus getLogUnitNodeStatusInLayout(Layout layout, String server) {
-                    // be established.
-                    cf.completeExceptionally(e);
+            // If a server is unresponsive attempt to ping for cluster_status_query_attempts
+            if(connectivityStatusMap.containsValue(ConnectivityStatus.UNRESPONSIVE)) {
+                // Ping only unresponsive endpoints
+                List<String> endpoints = connectivityStatusMap.entrySet()
+                        .stream()
+                        .filter(entry -> entry.getValue().equals(ClusterStatusReport.ConnectivityStatus.UNRESPONSIVE))
+                        .map(Entry::getKey)
+                        .collect(Collectors.toList());
+
+                for (String serverEndpoint : endpoints) {
+                    // Ping all nodes asynchronously
+                    CompletableFuture<Boolean> cf = runtimeLayout.getBaseClient(serverEndpoint).ping();
+                    pingFutureMap.put(serverEndpoint, cf);
                 }
-                futureMap.put(endpoint, cf);
+
+                // Accumulate all responses.
+                pingFutureMap.forEach((endpoint, pingFuture) -> {
+                    try {
+                        ClusterStatusReport.ConnectivityStatus connectivityStatus = CFUtils.getUninterruptibly(pingFuture,
+                                WrongEpochException.class) ? ConnectivityStatus.RESPONSIVE : ConnectivityStatus.UNRESPONSIVE;
+                        connectivityStatusMap.put(endpoint, connectivityStatus);
+                    } catch (WrongEpochException wee) {
+                        connectivityStatusMap.put(endpoint, ConnectivityStatus.RESPONSIVE);
+                    } catch (Exception e) {
+                        connectivityStatusMap.put(endpoint, ConnectivityStatus.UNRESPONSIVE);
+                    }
+                });
             }
+        }
 
-            // Accumulate all responses.
-            futureMap.forEach((endpoint, clusterStateFuture) -> {
-                try {
-                    clusterStatusMap.put(endpoint, CFUtils.getUninterruptibly(clusterStateFuture,
-                            WrongEpochException.class));
-                    counters.computeIfPresent(endpoint, (s, count) -> count + 1);
-                } catch (WrongEpochException wee) {
-                    counters.computeIfPresent(endpoint, (s, count) -> count + 1);
-                } catch (Exception ignored) {
-                    // Ignore all exceptions.
-                }
-            });
+        return connectivityStatusMap;
+    }
+
+    private Map<String, NodeStatus> getNodeStatusMap(Layout layout) {
+        Map<String, NodeStatus> nodeStatusMap = new HashMap<>();
+        for (String endpoint: layout.getAllServers()) {
+            if (layout.getUnresponsiveServers().contains(endpoint)) {
+                nodeStatusMap.put(endpoint, NodeStatus.DOWN);
+            } else if (layout.getSegments().size() != layout.getSegmentsForEndpoint(endpoint).size()) {
```

Comment:
This is assuming that all the nodes in the layout are LogUnit servers.
This should be fine for now. But can you add a comment noting this assumption here?

Reply:
@zalokhan I have some minor comments to add on other things, if it's ok I'll address this along those in a separate PR so we can merge this soon and generate a build.